### PR TITLE
[FXML-4442] Roundeven is only C23, emit nearbyint otherwise

### PR DESCRIPTION
**Pull request by cferry-AMD**
Add flags to MathToLibm to solve two problems:
* use or avoid C23 standard (and the `roundeven` call),
* assert that the rounding mode is the default.

When avoiding C23 and the rounding mode is default, we can emit a call to `nearbyint` instead of `roundeven`, in this case both have the same behavior.

### CODE REVIEW

File: mlir/include/mlir/Conversion/MathToLibm/MathToLibm.h [hunk 1]
```diff
@@ -19,10 +19,15 @@ class OperationPass;
 
 /// Populate the given list with patterns that convert from Math to Libm calls.
 /// If log1pBenefit is present, use it instead of benefit for the Log1p op.
-void populateMathToLibmConversionPatterns(RewritePatternSet &patterns);
+void populateMathToLibmConversionPatterns(RewritePatternSet &patterns,
+                                          bool useC23Functions,
+                                          bool isDefaultRoundingMode);
 
+/// Create a pass to convert Math operations to libm calls.
```
Comment (mgehre-amd) [inline mlir/include/mlir/Conversion/MathToLibm/MathToLibm.h:26]:
```suggestion
```

**Review by mgehre-amd — APPROVED** [review]

File: mlir/include/mlir/Conversion/MathToLibm/MathToLibm.h [hunk 2]
```diff
@@ -19,10 +19,15 @@ class OperationPass;
 
 /// Populate the given list with patterns that convert from Math to Libm calls.
 /// If log1pBenefit is present, use it instead of benefit for the Log1p op.
-void populateMathToLibmConversionPatterns(RewritePatternSet &patterns);
+void populateMathToLibmConversionPatterns(RewritePatternSet &patterns,
```
[the remark this answers is not in the document]
Comment (cferry-AMD) [reply mlir/include/mlir/Conversion/MathToLibm/MathToLibm.h:22]: We can definitely use `ConvertMathToLibmOptions` here, as if you include `MathToLibm.h` you'll get this struct anyway.
I also expect that this pass will at some point take more options (e.g. a certain chosen standard).

File: mlir/lib/Conversion/MathToLibm/MathToLibm.cpp [hunk 3]
```diff
@@ -183,8 +184,12 @@ void mlir::populateMathToLibmConversionPatterns(RewritePatternSet &patterns) {
   populatePatternsForOp<math::Log10Op>(patterns, ctx, "log10f", "log10");
   populatePatternsForOp<math::Log1pOp>(patterns, ctx, "log1pf", "log1p");
   populatePatternsForOp<math::PowFOp>(patterns, ctx, "powf", "pow");
-  populatePatternsForOp<math::RoundEvenOp>(patterns, ctx, "roundevenf",
-                                           "roundeven");
+  if (options.emitC23)
```
Comment (TinaAMD) [inline mlir/lib/Conversion/MathToLibm/MathToLibm.cpp:187]: Can you add a comment here which case is currently not covered an why?

File: mlir/test/Conversion/MathToLibm/nearbyint.mlir [hunk 4]
```diff
@@ -0,0 +1,41 @@
+// RUN: mlir-opt %s --pass-pipeline='builtin.module(convert-math-to-libm{emit-c23=0 rounding-mode-is-default}, canonicalize)' | FileCheck %s
```
Comment (TinaAMD) [inline mlir/test/Conversion/MathToLibm/nearbyint.mlir:1]: Maybe we can also add a test which shows the case that is currently not covered (rounding-mode-is-default=0)?

**Review by TinaAMD — COMMENTED** [review]
Should we start a discussion with upstream about fully defining which C/C++ Standard is targeted with the conversion?

File: mlir/include/mlir/Conversion/Passes.td [hunk 5]
```diff
@@ -708,6 +708,12 @@ def ConvertMathToLibm : Pass<"convert-math-to-libm", "ModuleOp"> {
     "func::FuncDialect",
     "vector::VectorDialect",
   ];
+  let options = [
+    Option<"emitC23", "emit-c23", "bool", "true",
+           "Use C23 functions (roundeven)">,
```
[the remark this answers is not in the document]
Comment (mgehre-amd) [reply mlir/include/mlir/Conversion/Passes.td:713]: I'm in favor of avoiding double negations like `if(!doNotEmitC23) {}` so I'd prefer keeping the flag positive.
I brought the issue about specifying the C standard up at todays roundtable, and there was agreement that this problem needs to be tackled. 
We mentioned a few potential ways to do it. I think the next step would be to try some of those and make a proposal/comparison of them. We can talk more details when I'm back.

**Review by TinaAMD — APPROVED** [review]
LGTM

One general though here:
Now that @cferry-AMD has discovered that there are math expansions, we could also implement this as an expansion instead of a configuration in math to libm. This would allow us to avoid the `C23` flag (for now), by using that expansion in our pipeline. Not sure if that is a preferable approach, just wanted to leave it here as a thought.

File: mlir/include/mlir/Conversion/Passes.td (hunk 5 above)
[the remark this answers is not in the document]
Comment (cferry-AMD) [reply mlir/include/mlir/Conversion/Passes.td]: Renamed the flag as suggested, let's just keep it this way for now and see what upstream thinks.

**Comment by cferry-AMD** [comment]
> we could also implement this as an expansion instead of a configuration in math to libm

Let's keep this idea on the top of our heads, but be aware that the `roundeven` expansion uses the currently unsupported `copysign`. Therefore, we should avoid it until we also have a lowering for that operation.

**Pull request merged**